### PR TITLE
Add statusCode and exceptionType Fields to Network Statsbeat

### PR DESCRIPTION
**Pull request by JacksonWeber**
This adds the statusCode and exceptionType fields to the appropriate network StatsBeat metrics per the spec https://github.com/microsoft/Telemetry-Collection-Spec/blob/main/ApplicationInsights/statsbeat.md

### CODE REVIEW

File: AutoCollection/Statsbeat.ts [hunk 1]
```diff
@@ -286,42 +306,45 @@ class Statsbeat {
...
-                this._statbeatMetrics.push({
-                    name: Constants.StatsbeatCounter.REQUEST_FAILURE,
-                    value: currentCounter.totalFailedRequestCount,
-                    properties: properties
+            if (currentCounter.totalFailedRequestCount.length > 0) {
+                currentCounter.totalFailedRequestCount.forEach((currentCounter) => {
+                    this._statbeatMetrics.push({
+                        name: Constants.StatsbeatCounter.REQUEST_FAILURE,
+                        value: currentCounter.count,
+                        properties: properties
```
Comment (hectorhdzg) [inline AutoCollection/Statsbeat.ts:311]: Where are you using currentCounter.statusCode?, I can see you have an accurate count but I'm not sure where are you actually adding it to the Statsbeat properties

File: Tests/AutoCollection/Statsbeat.tests.ts [hunk 2]
```diff
@@ -171,17 +171,17 @@ describe("AutoCollection/Statsbeat", () => {
...
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRetry(0, "test", 206);
+            statsBeat.countRetry(0, "test", 206);
+            statsBeat.countThrottle(0, "test", 206);
+            statsBeat.countException(0, "test", "Statsbeat Exception");
```
Comment (hectorhdzg) [inline Tests/AutoCollection/Statsbeat.tests.ts]: Add assert checking this exception type

File: Tests/AutoCollection/Statsbeat.tests.ts [hunk 3]
```diff
@@ -171,17 +171,17 @@ describe("AutoCollection/Statsbeat", () => {
...
-            statsBeat.countException(0, "test");
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, true, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRequest(0, "test", 1, false, 200);
+            statsBeat.countRetry(0, "test", 206);
+            statsBeat.countRetry(0, "test", 206);
```
Comment (hectorhdzg) [inline Tests/AutoCollection/Statsbeat.tests.ts:182]: Same here we want to validate you count multiple failed requests, retry and throttle and multiple Statsbeat are generated with the properties part of them

File: Tests/AutoCollection/Statsbeat.tests.ts [hunk 4]
```diff
@@ -171,17 +171,17 @@ describe("AutoCollection/Statsbeat", () => {
...
-            statsBeat.countRequest(0, "test", 1, true);
-            statsBeat.countRequest(0, "test", 1, true);
-            statsBeat.countRequest(0, "test", 1, false);
-            statsBeat.countRequest(0, "test", 1, false);
-            statsBeat.countRequest(0, "test", 1, false);
-            statsBeat.countRetry(0, "test");
-            statsBeat.countRetry(0, "test");
-            statsBeat.countThrottle(0, "test");
-            statsBeat.countException(0, "test");
+            statsBeat.countRequest(0, "test", 1, true, 200);
```
Comment (hectorhdzg) [inline Tests/AutoCollection/Statsbeat.tests.ts:174]: StatusCode should be optional parameter if is not used all the time, like in this case

Comment (JacksonWeber) [reply Tests/AutoCollection/Statsbeat.tests.ts:174]: If we make this parameter optional are we ok with creating entries in the request counter like `{ statusCode: undefined, count: <currentCount>`?

Comment (hectorhdzg) [reply Tests/AutoCollection/Statsbeat.tests.ts:174]: no need to do that, successful count is only a number

File: Tests/AutoCollection/Statsbeat.tests.ts (hunk 3 above)
Comment (JacksonWeber) [reply Tests/AutoCollection/Statsbeat.tests.ts:183]: Added some new tests that should ensure we're creating multiple metrics per statusCode now.

File: AutoCollection/Statsbeat.ts (hunk 1 above)
Comment (JacksonWeber) [reply AutoCollection/Statsbeat.ts:312]: Should be fixed now. Added it to the properties in each counter where it's relevant per the spec.

File: Tests/AutoCollection/Statsbeat.tests.ts (hunk 2 above)
Comment (JacksonWeber) [reply Tests/AutoCollection/Statsbeat.tests.ts]: Added tests to determine the exceptionType field is populated correctly on the metric.

File: AutoCollection/Statsbeat.ts [hunk 5]
```diff
@@ -118,43 +118,63 @@ class Statsbeat {
...
 
-    public countRequest(endpoint: number, host: string, duration: number, success: boolean) {
+    public countRequest(endpoint: number, host: string, duration: number, success: boolean, statusCode?: number) {
         if (!this.isEnabled()) {
             return;
         }
         let counter: Network.NetworkStatsbeat = this._getNetworkStatsbeatCounter(endpoint, host);
         counter.totalRequestCount++;
         counter.intervalRequestExecutionTime += duration;
         if (success === false) {
```
Comment (hectorhdzg) [inline AutoCollection/Statsbeat.ts:128]: Add check for statusCode here as well or you will create a null statusCode counter if this code path is ever triggered

**Review by hectorhdzg — APPROVED** [review]
LGTM

**Pull request merged**